### PR TITLE
feat: support dynamic templateref naming. Fixes: #10542

### LINKED ISSUE #10542: Support dynamic templateRef naming
# Summary

It would be great if we could use variables inside the `templateRef` field.

This feature will allow us to reuse a template with multiple implementations using inputs parameters.

# Use Cases

Below there is an example of what we're after: 

```yaml
templateRef:
  name: '{{inputs.parameters.workflow-template}}'
  template: '{{inputs.parameters.template-name}}'
```

Thanks!

---

**Message from the maintainers**:

Love this enhancement proposal? Give it a 👍. We prioritise the proposals with the most 👍.

### PR DESCRIPTION
**Pull request by shuangkun**
Support dynamic templateref naming

Fixes #10542

### CODE REVIEW

File: workflow/controller/operator_workflow_template_ref_test.go [hunk 1]
```diff
@@ -560,3 +560,77 @@ func TestSubmitWorkflowTemplateRefWithoutRBAC(t *testing.T) {
...
+            templateRef:
+              name: '{{item.workflow-template}}'
+              template: '{{item.template-name}}'
+            withItems:
+                - { workflow-template: 'hello-world-template-global-arg', template-name: 'hello-world'}
+          - name: hello-world-dag
+            template: diamond
+
+    - name: diamond
+      dag:
```
Comment (tczhao) [inline workflow/controller/operator_workflow_template_ref_test.go:604]: step and dag, nice

**Review by tczhao — COMMENTED** [review]
Seems the current code already resolves those templates just we never had them in the verify code, correct?

**Comment by shuangkun** [comment]
> Seems the current code already resolves those templates just we never had them in the verify code, correct?

Yes, "resolveReferences/resolveDependencyReferences" is execute before "expandStep/expandTask". So item will be resolved before executeTemplate.

**Review by tczhao — APPROVED** [review]
lgtm

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**